src/DispatchListener.php

```diff
@@ -61,11 +63,30 @@ public function attach(EventManagerInterface $events)
      */
     public function onDispatch(MvcEvent $e)
     {
-        $routeMatch       = $e->getRouteMatch();
+        $routeMatch     = $e->getRouteMatch();
+        $request        = $e->getRequest();
+        $application    = $e->getApplication();
+        $response       = $application->getResponse();
+        $serviceManager = $application->getServiceManager();
+
+        // middleware?
+        $middleware = $routeMatch->getParam('middleware', false);
+        if (false !== $middleware) {
+            if (is_string($middleware) && $serviceManager->has($middleware)) {
+                $middleware = $serviceManager->get($middleware);
+            }
+            if (!is_callable($middleware)) {
+                $middleware = is_string($middleware) ? $middleware : get_class($middleware);
+                $return = $this->marshalControllerNotFoundEvent($application::ERROR_MIDDLEWARE_CANNOT_DISPATCH, $middleware, $e, $application);
+                return $this->complete($return, $e);
+            }
+            $return = $middleware(Psr7Request::fromZend($request), Psr7Response::fromZend($response));
+            return $this->complete(Psr7Response::toZend($return), $e);
+        }
+
```

Comment:
I would argue we should separate this logic into a _different_ listener, which we would then add as a default listener in `Zend\Mvc\Application`; it would execute at higher priority than this one (or same priority, but register prior to this one). If no middleware route match parameter is found, it would simply return, which would cause the original dispatch listener to take over.